### PR TITLE
QUEUE- and Unix-backend-related improvements, bugfixes, and tests. Version 4

### PR DESCRIPTION
**Pull request by ahippo**
This is an improved version of #577, which instead of exposing the whole queue.h to a user and making it a part of the public API, exposes only the absolutely necessary part of queue.h (`struct QUEUE` declaration).
This pull request also incorporates comments raised as part of review request #577.
Everything else is identical to #577.

I didn't force-push new changesets to #577 as this would wipe many comments there because they are tied to particular changesets.

### CODE REVIEW

**Comment by ahippo** [comment]
@saghul, since this PR is marked as v2, should I rebase it on top of master instead of v1.x?
Then I can also include commit cb5d8d732eead4cda1f9e289fe05eedcecdf06ba (which breaks ABI) into the PR.

**Comment by txdv** [comment]
O yeah, I remember this PR. It was really awesome, the inline instead of the do while seems a modern approach to do while preprocessor magic. Hopefully @saghul has more time to review it this time!

**Comment by saghul** [comment]
@ahippo doh, this fell through the cracks, I'm afraid! Yeah, please rebase on top of master since we're now getting started with v2 stuff.

**Comment by ahippo** [comment]
> O yeah, I remember this PR. It was really awesome, the inline instead of the do while seems a modern approach to do while preprocessor magic. Hopefully @saghul has more time to review it this time!

Thank you!
At least, it seemed more debug-able to me in case of a queue corruption.

> @ahippo doh, this fell through the cracks, I'm afraid! Yeah, please rebase on top of master since we're now getting started with v2 stuff.

No worries, it was quite intrusive for a stable branch.
Ok, I'll rebase on top of master and update the pull request.

**Comment by ahippo** [comment]
I've just rebased my changes on top of master branch.
My branch is now confusingly called v1.x-uv-queue-v4 although changes are meant for master not v1.x.
(I kept the old branch name in order to reuse this pull request)

Comments are welcome!
Thank you in advance!
Please, let me know if you have any questions.

**Comment by cjihrig** [comment]
@ahippo are you interested in picking this back up? It has a number of conflicts at the moment.

**Comment by ahippo** [comment]
> @ahippo are you interested in picking this back up? It has a number of conflicts at the moment.

@cjihrig, yes, I'm still interested.
Should I rebase it on top of the current master?

**Comment by cjihrig** [comment]
Yes, I think that's a good place to start.

**Comment by ahippo** [comment]
> Yes, I think that's a good place to start.

OK, I'll work on rebasing the PR on top of current master, and will update the PR.

File: include/uv/queue.h [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+ * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ */
+
+#ifndef UV_QUEUE_H_
+#define UV_QUEUE_H_
+
+#ifdef __cplusplus
```
Comment (cjihrig) [inline include/uv/queue.h]: Are these blocks necessary?

File: src/queue.h [hunk 2]
```diff
@@ -34,75 +37,90 @@ typedef void *QUEUE[2];
...
+ * should be done using QUEUE_REMOVE_SAFE() in conjunction with
+ * QUEUE_FOREACH_SAFE() sharing the same n.
+ * It's still not safe to remove the head though.
+ */
+#define QUEUE_FOREACH_SAFE(q, n, h)                                           \
+  for ((q) = (h)->next, (n) = (q)->next;                                      \
+       (q) != (h);                                                            \
+       (q) = (n), (n) = (n)->next)
+
+static inline int QUEUE_EMPTY(const QUEUE *q) {
```
Comment (cjihrig) [inline src/queue.h]: Can you align the pointer stars to the left throughout.

Comment (ahippo) [reply src/queue.h]: Sure thing.
I'll post an updated version after gathering some more comments.

**Comment by ahippo** [comment]
> Left a couple comments. To be honest, I'm still not entirely sure what the end goal of this PR is. Would it make more sense to split it up into smaller, more easily digestible PRs?

GitHub shows commits in a wrong order, which makes it even harder to understand what's going on.
I can definitely split the PR into multiple PRs if that would help with the review.
Is this what you want me to do or first wait for others to review the PR as is?

There are four goals overall:

1. Convert queue implementation to a struct and inline functions to make debugging and maintenance easier. This also makes the related code a bit more self-documented.
1. Increase coverage of a queue test.
1. Use pretty standard `QUEUE_FOREACH_SAFE()` for iterating over a queue while possibly removing elements from it.
1. Use QUEUE and related machinery in several places replacing an open-coded implementation.

**Comment by bnoordhuis** [comment]
Closing because this PR is unfortunately very stale.

**Pull request closed without merging**